File: 03-poo/15-diseno-orientado-a-objetos/java/ejemplos/Responsabilidades.java
public class Responsabilidades {
    static class ValidadorUsuario { boolean esValido(String correo){ return correo != null && correo.contains("@"); } }
    static class Notificador { void enviar(String correo){ System.out.println("Notificación a " + correo); } }
    static class RegistroUsuario {
        private final ValidadorUsuario validador = new ValidadorUsuario();
        private final Notificador notificador = new Notificador();
        void registrar(String correo){ if(!validador.esValido(correo)) throw new IllegalArgumentException(); notificador.enviar(correo); }
    }
    public static void main(String[] args){ new RegistroUsuario().registrar("aprendiz@example.com"); }
}

File: 03-poo/15-diseno-orientado-a-objetos/scala/ejemplos/Responsabilidades.scala
class ValidadorUsuario:
  def esValido(correo: String): Boolean = correo != null && correo.contains("@")
trait Notificador:
  def enviar(correo: String): Unit
class ConsolaNotificador extends Notificador:
  def enviar(correo: String): Unit = println(s"Notificación a $correo")
class RegistroUsuario(validador: ValidadorUsuario, notificador: Notificador):
  def registrar(correo: String): Unit =
    require(validador.esValido(correo))
    notificador.enviar(correo)

@main def responsabilidades(): Unit = new RegistroUsuario(new ValidadorUsuario,new ConsolaNotificador).registrar("aprendiz@example.com")
